copt/randomized.py

```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
             memory_gradient[i] = grad_i
 
     return epoch_iteration_template
+
+
+def step_size_sfw(variant):
+    if variant in {'SAG', 'SAGA'} :
+        def step_sizes_SAG_A(t, n_samples=None):
+            step_size_x = 2 / (t+2)
+            return step_size_x, None
+        return step_sizes_SAG_A
+
+    if variant == 'MK':
+        def step_sizes_MK(t, n_samples=None):
+            step_size_x = 1 / (t+1)
+            step_size_agg = step_size_x ** (2/3)
+            return step_size_x, step_size_agg
+        return step_sizes_MK
+
+    if variant == 'LF':
+        def step_sizes_LF(t, n_samples=None):
+            if n_samples is None:
+                raise ValueError("n_samples must be the number of samples in the dataset.")
+            step_agg = 2 * n_samples / (2 * n_samples + t + 1)
+            step_x = 2 * (2 * n_samples + t) / ((t+1) * (4 * n_samples + t + 1))
+            return step_x, step_agg
+        return step_sizes_LF
+
+
+def minimize_sfw(
+        f_deriv,
+        A,
+        b,
+        x0,
+        lmo,
+        step_size=None,
+        max_iter=500,
+        tol=1e-6,
+        verbose=False,
+        callback=None,
+        variant='SAGA'
+):
+    r"""Stochastic Frank-Wolfe (SFW) algorithm.
+
+    The SFW algorithm can solve optimization problems of the form
```

Comment:
I think this is a bit biased. I think that Mokhtari and other papers in that line of work are more general and can be applied to streams, for example.

Reply:
True, I'll note that it is for this implementation.